gigascience/gigadb-website
# Email sent to author upon "DataPending" status needs to be customisable

# Pull request for issue: #1618

This is a pull request for the following functionalities:

* If editor sets status to DataPending, they are going to see a modal popup that prompts them for the email sent to the author

## How to test?

* Go to edit any dataset entry
* Set status from, e.g. "submitted" to "DataPending" and save, you are expected to see a modal popup (you might need to follow the fuw workflow):
![image](https://github.com/gigascience/gigadb-website/assets/143437854/a33de1b8-5b21-4ad9-ac37-dab590f26297)
* Validate appearance and copy of modal content
* Navigate modal with keyboard
* Update textarea value to anything you want, optionally including `{{ identifier }}`
* Click on "save and send email"
* Verify that a new email with the expected content appears in `fuw/app/frontend/runtime/mail`

## How have functionalities been implemented?

* Different options to implement this feature were evaluated, e.g. dynamically display additional field under upload status field, display persistent field, ... I opted for the current solution because it does not bloat the form, there's no risk that editor forgets about it or misses it and to me it makes sense that from a UI perspective it's not part of the form because it's not a dataset metadata or attribute (but technically the email textarea is part of the form and submitted as formdata along with the rest).
* The "save" button of the form conditionally opens a modal (instead of submitting the form) if the editor has changed the upload status to "DataPending" (from some other value)
* The modal shows an additional input that alllows the editor to customize the email sent to the author. this field is required which means that the editor cannot submit the form with an empty value
* If the "save and send" button in the modal is clicked, the form is submitted with the additional input value
* the controller intercepts the post request. If the formdata is in the "datapending" case and there is an email message defined, the controller will use this email message, parse the identifier placeholder, and submit as email to the author. If this email message is not defined, the controller defaults to the usual behavior (using a twig template)
* the default value for the email message is copied from the twig template and is hardcoded within the script tag
* IMPORTANT: if in production some other twig template is used, this value needs to be updated to whatever value it has in production. Alternatively I can investigate how to get the text from the twig template and show it in the frontend

## Any issues with implementation?

--

## Any changes to automated tests?

--

## Any changes to documentation?

--

## Any technical debt repayment?

--

## Any improvements to CI/CD pipeline?

--
